File: runtime/peerbot/src/com/newbound/p2p/protocol/RESPONSE.java
package com.newbound.p2p.protocol;

import javax.crypto.BadPaddingException;

import org.json.JSONObject;

import com.newbound.crypto.SuperSimpleCipher;
import com.newbound.p2p.P2PPeer;
import com.newbound.p2p.P2PService;
import com.newbound.p2p.P2PSocket;
import com.newbound.robot.BotUtil;
import com.newbound.robot.Callback;

public class RESPONSE implements Callback 
{

	P2PService P2P;

	public RESPONSE(P2PService p2p) 
	{
		super();
		P2P = p2p;
	}

	@Override
	public void execute(final JSONObject data) 
	{
		final Callback me = this;
		
		P2P.CONTAINER.getDefault().addJob(new Runnable() 
		{
			@Override
			public void run() 
			{
				try
				{
					String uuid = data.getString("uuid");
					byte[] ba = (byte[])data.get("data");
					P2PSocket sock = (P2PSocket)data.get("p2psocket");
					try
					{
						P2PPeer p = P2P.getPeer(uuid);
//						p.setLastContact(System.currentTimeMillis());

						ba = p.decrypt(ba, 0, ba.length);

						p.response(ba);
					}
					catch (BadPaddingException x)
					{
						try { P2P.getPeer(uuid).setReadKey((SuperSimpleCipher)null); } catch (Exception x2) {x2.printStackTrace();}
						P2P.isOK(uuid, sock, me, data);
					}
					catch (Exception x) { x.printStackTrace(); }
				}
				catch (Exception x) { x.printStackTrace(); }
			}
		}, "Executing P2P command");
	}
	

}

File: runtime/peerbot/src/newbound/p2p/protocol/response.py
class RESPONSE(object):
    def __init__(self, service):
        self.service = service

    def execute(self, data):
        def doit():
            uuid = data['uuid']
            ba = data['data']
            sock = data['p2psocket']
            p = self.service.p2p.getPeer(uuid)
            ba = p.decrypt(ba)
            try:
                p.response(ba)
            except Exception as e:
                print('Error receiving p2p response '+str(e))
                self.service.p2p.printStackTrace(e)

        self.service.p2p.addJob(doit, 'Receiving P2P response')
